main/src/main/scala/sbt/internal/RunUtil.scala

```diff
@@ -0,0 +1,301 @@
+    )
+    RunInfo(
+      jvm = true,
+      jvmRunInfo = jvmRunInfo,
+    )
+
+  def defaultRunMainTask(
+      products: Initialize[Task[Classpath]],
+      classpath: Initialize[Task[Classpath]],
+      scalaRun: Initialize[Task[ScalaRun]],
+      clientRun: Initialize[Boolean],
+  ): Initialize[InputTask[Unit | ClientJobParams]] =
+    val parser = Defaults.loadForParser(discoveredMainClasses)((s, names) =>
+      Defaults.runMainParser(s, names getOrElse Nil)
+    )
+    Def.inputTask {
+      val conv = fileConverter.value
+      given FileConverter = conv
+      val service = bgJobService.value
+      val (mainClass, args) = parser.parsed
+      val hashClasspath = (bgRunMain / bgHashClasspath).value
+      val fo = (run / forkOptions).value
+      val state = Keys.state.value
+      if clientRun.value && state.isNetworkCommand then
+        val workingDir = service.createWorkingDirectory
+        val cp = service.copyClasspath(
+          products.value,
+          classpath.value,
+          workingDir,
+          conv,
+        )
+        val info = mkRunInfo(args.toVector, mainClass, cp, fo, conv)
+        val result = ClientJobParams(
+          runInfo = info
+        )
+        import sbt.internal.worker.codec.JsonProtocol.*
+        state.notifyEvent(Serialization.clientJob, result)
+        result
+      else
+        val wrapper = termWrapper(canonicalInput.value, echoInput.value)
+        val handle = service.runInBackgroundWithLoader(Keys.resolvedScoped.value, state):
+          (logger, workingDir) =>
+            val cp = service.copyClasspath(
+              products.value,
+              classpath.value,
+              workingDir,
+              hashClasspath,
+              conv,
+            )
+            scalaRun.value match
+              case r: Run =>
+                val loader = r.newLoader(cp.files)
+                (
+                  Some(loader),
+                  wrapper(() => r.runWithLoader(loader, cp.files, mainClass, args, logger).get)
+                )
+              case sr =>
+                (None, wrapper(() => sr.run(mainClass, cp.files, args, logger).get))
+        service.waitForTry(handle).get
+        ()
+    }
+
+  def defaultRunTask(
+      products: Initialize[Task[Classpath]],
+      classpath: Initialize[Task[Classpath]],
+      mainClassTask: Initialize[Task[Option[String]]],
+      scalaRun: Initialize[Task[ScalaRun]],
+      clientRun: Initialize[Boolean]
+  ): Initialize[InputTask[Unit | ClientJobParams]] =
+    val parser = Def.spaceDelimited()
+    Def.inputTask {
+      val conv = fileConverter.value
+      given FileConverter = conv
+      val args = parser.parsed
+      val service = bgJobService.value
+      val mainClass = getMainClass(mainClassTask.value)
+      val hashClasspath = (bgRun / bgHashClasspath).value
+      val fo = (run / forkOptions).value
+      val state = Keys.state.value
+      if clientRun.value && state.isNetworkCommand then
```

Comment:
As suggested by @adpi2, this moves the condition logic into `run` input task, which allows plugins to opt out of this behavior.